File: algorithms/147. Insertion Sort List/solution.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def insertionSortList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if not head:
            return
        dummy = ListNode(None)
        cur = dummy.next = head
        while cur.next:
            if cur.next.val < cur.val:
                p = dummy
                while p.next.val < cur.next.val:
                    p = p.next
                t = cur.next
                cur.next = t.next
                t.next = p.next
                p.next = t
            else:
                cur = cur.next
        return dummy.next

File: algorithms/147. Insertion Sort List/solution.cc
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* insertionSortList(ListNode* head) {
        ListNode dummy(INT_MIN);
        dummy.next = head;
        ListNode * cur = head, * tmp;
        while ( cur && cur->next ) {
            if ( cur->val > cur->next->val ) {
                head = & dummy;
                while ( head->next->val <= cur->next->val ) {
                    head = head->next;
                }
                tmp = head->next;
                head->next = cur->next;
                cur->next = cur->next->next;
                head->next->next = tmp;
            } else {
                cur = cur->next;
            }
        }
        return dummy.next;
    }
};
